ROOT-198 block search form autocomplete

jesconstantine opened the pull request:
This PR begins implementing autocomplete functionality for the block search form:

- stubs out a view rest export providing an endpoint to search content by the title and render an array of item objects with title + url (whose keys reflect their solr schema counterpoints)
- adds a module library with css + js for autocomplete functionality, which hooks into the block search form

## @TODO
- [x] Add block configuration options which will get passed to JS (either in drupal settings or on the block itself)
	- enable autocomplete
    - endpoint (url or view)
	- number of results
	- results title text
	- keyboard directions visibility
	- number of characters after which to trigger the search
	- soon: mode?
- [x] Conditionally attach libraries in block form build() based on being enabled
- [x] Handle settings with smart defaults in the js
- [x] Document the expected result structure
- [x] Test view with "search terms" vocab on nodes (GTA's D7 views logic was searching title OR synonyms)

agentrickard committed:
Test template file

nstriedinger committed:
Update search_api_federated_solr.module

nstriedinger committed:
Update FederatedSearchPageFormBlock.php

nstriedinger committed:
Update federated-search-page-form-block.html.twig

agentrickard committed:
Adds proper template files for block and block form.

agentrickard committed:
Updates theme doc blocks

agentrickard committed:
Updates documentation for theming.

agentrickard committed:
Minor docs update

agentrickard committed:
Fixes typo

jesconstantine committed:
root-198: stub out autocomplete functionality for search block

jesconstantine committed:
root-198: use Drupal.announce, highlight results, use better logic

jesconstantine committed:
root-198: use view rest export

jesconstantine committed:
root-198: adhere to solr schema for title/url

jesconstantine committed:
root-198: add stubbed out view rest export for autocomplete

jesconstantine committed:
Root-198: add js hook to search form

agentrickard commented:
Notes:

* Shouldn't the title filter actually be a Contextual Argument, making the url http://d8.fs-demo.local/search-api-federated-solr-block-form-autocomplete-rest/soup?
* The current filter has the label "term" which makes the callback http://d8.fs-demo.local/search-api-federated-solr-block-form-autocomplete-rest?term=soup
* I believe the view would go in /config/optional and not /config/install

agentrickard commented:
And here's code for getting the list of Views for the config.

https://github.com/agentrickard/workbench/blob/8.x-1.x/src/Form/WorkbenchSettingsForm.php#L75

And then retrieval and usage:

https://github.com/agentrickard/workbench/blob/8.x-1.x/src/Controller/WorkbenchContentController.php#L175 etc.

jesconstantine committed:
root-198: scaffold out federated search page block config

jesconstantine committed:
root-198: make config available in drupalSettings for block

jesconstantine committed:
root-198: begin working with block config in js

jesconstantine committed:
root-198: create view rest export serializer to structure solr response

jesconstantine committed:
root-198: move autocomplete view to optional, use new serializer

jesconstantine committed:
root-198: fix config item paths in bock config form

jesconstantine committed:
root-198: update js response handler to expect solr response structure

jesconstantine committed:
root-198: get url from config, append default value params, wildecard

jesconstantine committed:
root-198: append expected params to default url

jesconstantine committed:
root-198: don't make module depend on rest (since view is optional config)

jesconstantine committed:
root-198: only expose block form autocomplete config when enabled

jesconstantine committed:
root-198: support basic auth in ajax request

jesconstantine committed:
root-198: honor the default values set in JS

jesconstantine committed:
root-198: Fix bug where autocomplete is enabled was not being written to config

jesconstantine commented:
@agentrickard thanks for your feedback.  In regards to your questions:

- What I liked about using exposed filters was the flexibility with multiple filter fields.  For example, if we need to search multiple fields (i.e. the title and the search terms taxonomy ref field) we have that power, we can set the logic to OR or AND, and we can set the param key to be whatever we want (and we can update our URL in the block autocomplete config respectively).
  Is this still possible with contextual filters?  I was never able to get up and running very well going that route.  I have update the view to include a display with title as contextual filter though.  I'll try to touch base with someone to help me out.
- Yes, the filter identifier was intentional, I've updated the logic and documentation for the URL field of the block config.  Let me know what you think?
    ![image](https://user-images.githubusercontent.com/3279883/54151050-4fc53300-4410-11e9-9b09-f0cc2a48aea2.png)
- Thanks, I've moved the view config to `config/optional` and removed `rest` as a module dependency (yes!).

Other updates since your last review include:

1. Add support for wildcard searches
1. Add support for basic auth (and remembered why there is a separate field in the search app, its because the creds for the app should be read only while the server will require read + write)
1. Created a views (rest export) style plugin to serialize the result structure, matching the structure of the solr response object.  That way if folks create a view (of either content or search api index) to use for autocomplete, so long as they create a rest export that returns json, and use the proper alias for the exposed title + link fields then these will work with the js.
    -  search api view using solr_serializer:
        ![image](https://user-images.githubusercontent.com/3279883/54151740-e34b3380-4411-11e9-82ad-98c9028771b6.png)
    - content view using solr_serializer:
        ![image](https://user-images.githubusercontent.com/3279883/54151789-04ac1f80-4412-11e9-905d-655b660f79b2.png)


I think with this new config layout, we don't care if they use a view / solr / something else, as long as they provide us with the URL and all of its params.  So I didn't think it necessary to implement any of the views logic at this point.

Next steps:
- [x] Update the module with the patch to include the theming changes and make any necessary changes (might need to update a selector or two in the js)
- [x] Attach the autocomplete JS library conditionally in the block form front end (right now its attached unconditionally)
- [x] Update the docs for the Autocomplete URL in the app settings form and make the show/hide behavior for autocomplete in the app config match that of the block config.
- [x] Test autocomplete / view / boosting with search term tax ref field in demo site

Anything else you can think of?

agentrickard commented:
The use of filter makes sense now, because Contextual Filters cannot handle an unknown set of qualifiers. Don't bother changing it; just document why we use Exposed Filters.

Double-check that `optional` views get installed. I think they do, but I don't think they trigger requirements checks. I'm not entirely sure. 

It would be even more ideal if we could alias the field names `ss_federated_title` and `ss_url` at the app level so that we recognized `title` and `url` as the required fields and just auto mapped our unique names to those values.

Can you only enter one `[val]` in the URL? Is that restriction enforced?

jesconstantine commented:
@agentrickard optional config only gets installed if its dependencies are installed.  So if the following modules are installed in the site when they install our module, the view will be installed as well: node, rest, serialization, user

re: the field name aliases, I kept it that way because my understanding was that we wanted to enforce a result structure which mapped to our index schema.  The serializer + field aliases will help with that.

However, I do have a todo comment which is: support a custom transform function at the app level.  Basically, the app would look for `drupalSettings.searchApiFederatedSolr.block.autocomplete.transform` and if it exists and if it's a function, would execute it on the results returned from the ajax request before rendering them.  That way it doesn't matter what the structure returned from the passed in URL is, the structure that we'd iterate over before rendering would always be like the solr response object.

Then we'd document that implementers create a library that is a JS function writing the contents of their function to drupalSettings, load their library in the head (or some weight that we recommend?) and preprocess the block, attaching their library.

That seemed like a little much for right now, though.  What do you think?

re: multiple instances of `[val]` we will need to support multiple instances (for example if they use a view and filter across multiple fields we'd need to support `title=[val]&search_terms=[val]` right?.  I'll update the JS to use a regex with `/g` as the arg that gets passed to `String.prototype.replace()` to be sure this works.

jesconstantine committed:
root-198: replace multiple instances of [val] in autocomplete URL

jesconstantine committed:
root-198: allow autocomplete to be disabled <smh>

jesconstantine committed:
root-198: attach library + drupal settings to block if autocomplete enabled

jesconstantine committed:
root-198: provide better namespace for autocomplete in drupalSettings

jesconstantine committed:
root-198: remove contextual filter display from example view

agentrickard commented:
Its a little much right now.

Regarding transformation, I was more thinking of cases where the results didn't come from Drupal. We can leave it as is for now.

agentrickard commented:
AFAIK: The YML file for the view cannot have a UUID if it is distributed as part of the module.

agentrickard commented on src/Form/FederatedSearchPageBlockForm.php:
```diff
@@ -41,6 +41,8 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#attributes' => [
         'title' => $this->t('Enter the terms you wish to search for.'),
         'placeholder' => 'Search',
+        'autocomplete' => "off",
```
This needs to read from config:

`        'autocomplete' => $app_config->get('autocomplete.isEnabled') ? 'on' : 'off',`

agentrickard replied:
Don't we also have to attach the JS library here?

jesconstantine replied:
@agentrickard good questions.

re the autocomplete attribute, this is actually just disabling the [html attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) which allows the user agent to make suggestions based on browser history and can interfere with our autocomplete functionality.

Our autocomplete trigger class, library and drupal settings are attached to the block, based on block config in https://github.com/palantirnet/search_api_federated_solr/pull/78/commits/5269923de949babbb23d26e2f78c2d0718265352

jesconstantine committed:
root-198: add the js trigger class to the block based on config

jesconstantine committed:
root-198: add comment to clarify autocomplete html attribute

jesconstantine committed:
root-198: remove uuid from autocomplete view

jesconstantine committed:
root-198: update search app autocomplete url docs + other misc fixes

- make autocomplete enable checkbox label clickable
- open/close autocomplete details based on dis/enabled
- make enable checkbox trigger for other form values, not whole details

jesconstantine committed:
root-198: emit event when a search block autocomplete option is selected

jesconstantine committed:
root-198: add custom event emission to autocomplete link click

agentrickard commented:
I see. I was failing to configure the block properly. We should link to that in the search form -- or move the block config to the search app config page.

jesconstantine commented:
@agentrickard I know I can link to the search app settings page from the block config form, but I didn't know how to link to the block config form from the search app settings because I thought that route would be determined by the block instance title which is user editable.  Is that accurate?

I think it will be less of an issue for new module adopters because they'll see the config form when they place the block.  And this afternoon I plan to update all of the module docs with instructions for autocomplete:

- new options in the app settings
- new options in the block config
- implications of solr config
- implications of full text fields in the index

agentrickard commented:
Works for me.

nstriedinger committed:
add suggestions and update templates

jesconstantine committed:
root-198: use custom data attr (vs name) as hook to show/hide form elements

Fixes bug where all of the sudden the block config enable autocomplete value was not being honored

jesconstantine committed:
root-198: set install, schema, + default values for autocomplete config

jesconstantine commented:
@agentrickard I think I've updated the rest of the install config + the schema for the module:
- https://github.com/palantirnet/search_api_federated_solr/pull/78/files#diff-30dc0e17facd70f3dc4aa6dea143af28
- https://github.com/palantirnet/search_api_federated_solr/pull/78/files#diff-3a7c4e98d40d452abf8430d5d5148024

I've also created an update hook to set the default config values: 
- https://github.com/palantirnet/search_api_federated_solr/pull/78/files#diff-5d98e2ca694609ca7a9436eed4c3f6d0

Can you remind me of the best way to test all of this?

jesconstantine committed:
Merge branch '3037751-template' into root-198

agentrickard commented:
The best way to test would be to install an older version of the module. Get it working, and then update the code and see that it still works.

That's hard to do in an automated test, though you could Kernel Test it, I think, by setting the old config, then running the update, then checking the config. Kernel tests, IIRC, do not auto-install config.

jesconstantine committed:
root-198: fix readme, make numbered list item links render as links

jesconstantine committed:
root-198: fix readme: point required schema docs to d.o guide

jesconstantine committed:
root-198: update readme: include steps to set default query fields, block config

jesconstantine committed:
root-198: update readme with autocomplete config, fields, views docs

jesconstantine commented:
@agentrickard I've begun updating the readme with some documentation: https://github.com/palantirnet/search_api_federated_solr/pull/78/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8

To test the config/schema/update hook I took these steps and they appeared to have worked:

### To test the update hook
1. Checked out the `8.x-1.x` version of the module in `src/search_api_federated_solr`
1. Checked out the `root-198-search-terms-field` in the project root for the demo site (which successfully adds a search terms field to recipes, articles as a default field in the solr server config, and as full text field to the index)
1. Reprovision the vm with `vagrant provision` to get the updated solr config
1. Rebuild the D8 site from within the project root on the vm: `phing install-d8 -Dbuild.env=d8`
1. Reinitialize the modules in src `phing init`
1. You should be at the last released state of the module (`drush @fsd-d8.local uli` and poke around to confirm all autocomplete work is gone)
1. Check out this branch from within `src/search_api_federated_solr`
1. Rebuild cache `drush @fsd-d8.local cr`
1. Run the update hook for the module `drush @fsd-d8.local updb`
1. The update hook should be recognized, execute it, it should finish successfully
1. On the site, your search form and search app should still function as it did, but now you should have the autocomplete config added to the block config and the search app config
1. Test a bit of the configuration to ensure that autocomplete functionality works (use the default endpoints for now)

### To test the install config (including the view)
This felt a bit hacky but it also worked, I think

1. Uninstall the module `drush @fsd-d8.local pmu -y search_api_federated_solr`
1. Remove any trace of it from your local config in `config/sites/d8`:
    - disable the `search_api_federated_solr` module in core.extensions
    - block.block.federatedsearchpageformblock
	- block.block.stark.federatedsearchpageformblock
	- image.style.search_api_federated_solr_image.yml
	- search_api_federated_solr.search_app.settings
1. Import config `drush @fsd-d9.local cim -y`
1. Install the module `drush @fsd-d8.local en -y search_api_federated_solr`
1. The view should have installed
1. Place the block in the highlight region
1. Configure the block and app

Is there anything else I'm missing?

agentrickard commented on README.md:
```diff
@@ -74,7 +76,11 @@ In order to display results from the Solr index:
 1. Optional: [Theme the ReactJS search app](docs/theme.md)
 1. Optional: Add the federated search page form block to your site theme + configure the block settings
-
+1. Optional: If you want autocomplete functionality and would prefer that results come from a view, [create a Search API search view with a rest export](https://www.drupal.org/docs/8/modules/search-api/getting-started/search-forms-and-results-pages/searching-with-views-0) or create a content view with a rest export (see view added as optional config for this module in `config/optional`) and use t ...
```
I would include teh name of our provided View

jesconstantine replied:
Thanks, added in https://github.com/palantirnet/search_api_federated_solr/pull/78/commits/cc76a5643267dfb627b775ac217ab61b96c3987b

agentrickard commented:
I don't think so, test-wise.

jesconstantine committed:
root-198: document autocomplete view name in readme

jesconstantine commented:
@agentrickard are you comfortable with me merging this into the root-198-autocomplete integration branch, with the understanding that it'll all be tested again once dependencies are updated?

agentrickard commented:
Yes.
